types/transactions.go

```diff
@@ -143,17 +144,51 @@ type (
 // ID returns the id of a transaction, which is taken by marshalling all of the
 // fields except for the signatures and taking the hash of the result.
 func (t Transaction) ID() TransactionID {
-	return TransactionID(crypto.HashAll(
-		t.SiacoinInputs,
-		t.SiacoinOutputs,
-		t.FileContracts,
-		t.FileContractRevisions,
-		t.StorageProofs,
-		t.SiafundInputs,
-		t.SiafundOutputs,
-		t.MinerFees,
-		t.ArbitraryData,
-	))
+	var hash crypto.Hash
+	h := crypto.NewHash()
+	enc := encoding.NewEncoder(h)
+
+	for i := range t.SiacoinInputs {
+		t.SiacoinInputs[i].MarshalSia(h)
+	}
+	encoding.WriteInt(h, len((t.SiacoinOutputs)))
+	for i := range t.SiacoinOutputs {
+		t.SiacoinOutputs[i].MarshalSia(h)
+	}
+	encoding.WriteInt(h, len((t.FileContracts)))
+	for i := range t.FileContracts {
+		enc.Encode(t.FileContracts[i])
+	}
+	encoding.WriteInt(h, len((t.FileContractRevisions)))
+	for i := range t.FileContractRevisions {
+		enc.Encode(t.FileContractRevisions[i])
+	}
+	encoding.WriteInt(h, len((t.StorageProofs)))
+	for i := range t.StorageProofs {
+		enc.Encode(t.StorageProofs[i])
+	}
+	encoding.WriteInt(h, len((t.SiafundInputs)))
+	for i := range t.SiafundInputs {
+		enc.Encode(t.SiafundInputs[i])
+	}
+	encoding.WriteInt(h, len((t.SiafundOutputs)))
+	for i := range t.SiafundOutputs {
+		t.SiafundOutputs[i].MarshalSia(h)
+	}
+	encoding.WriteInt(h, len((t.MinerFees)))
+	for i := range t.MinerFees {
+		t.MinerFees[i].MarshalSia(h)
+	}
+	encoding.WriteInt(h, len((t.ArbitraryData)))
+	for i := range t.ArbitraryData {
+		encoding.WritePrefix(h, t.ArbitraryData[i])
+	}
+	encoding.WriteInt(h, len((t.TransactionSignatures)))
+	for i := range t.TransactionSignatures {
```

Comment:
You can see from the asm that in most cases it doesn't actually allocate or copy anything though. The compiler sees that it can safely use the existing memory. The only reason there's a performance difference is that it computes the memory offset less efficiently (unclear why).

Regardless, clarity comes first, so we should use the value form unless the index is needed or (as in your example) you need a pointer to the value. It's less verbose and communicates the intent more clearly.